nsv13/code/modules/collision/shape.dm

```diff
@@ -0,0 +1,141 @@
+
+/*
+Method to set our position to a new one.
+*/
+
+/datum/shape/proc/_set(_x, _y)
+	position.x = _x
+	position.y = _y
+
+/*
+Method to set our points to a new list of points
+@param points = a list filled with pixel point coordinates
+*/
+
+/datum/shape/proc/set_points(list/points)
+	src.points = points
+	if(!src.base_points || src.base_points.len != points.len){
+		src.rel_points.Cut()
+		src.normals.Cut()
+		for (var/i in 1 to points.len){
+			src.rel_points.Add(new /datum/vector2d(0,0))
+			src.normals.Add(new /datum/vector2d(0,0))
+		}
+		src.base_points = points
+		src._recalc()
+	}
+	return points
+
+/*
+Method to set our angle to a new angle as required, then recalculate our points as necessary.
+*/
+/datum/shape/proc/set_angle(var/angle)
+	if(angle == src._angle){ //No need for expensive recalculation for very minor changes.
+		return FALSE
+	}
+	src._angle = angle
+	src._recalc()
+
+/*
+Method to recalculate our bounding box, adjusting the relative positions accordingly
+*/
+
+/datum/shape/proc/_recalc()
+	for(var/i in 1 to src.base_points.len){
+		src.rel_points[i]._set(src.base_points[i].rotate(src._angle))
+	}
+	//Clear out our current AABB collision box
+	src.aabb.Cut()
+	var/min_x = INFINITY
+	var/min_y = INFINITY
+	var/max_x = -INFINITY
+	var/max_y = -INFINITY
+	//Recalculate the points
+	for(var/i in 1 to src.rel_points.len){
+		var/datum/vector2d/p1 = src.rel_points[i]
+		var/datum/vector2d/p2 = i < src.base_points.len ? src.rel_points[i+1] : src.rel_points[0]
+
+		if(p1.x < min_x){
+			min_x = p1.x
+		}
+		if(p1.y < min_y){
+			min_y = p1.y
+		}
+		if(p1.x > max_x){
+			max_x = p1.x
+		}
+		if(p1.y > max_y){
+			max_y = p1.y
+		}
+
+		var/datum/vector2d/edge = p2 - p1
+		src.normals[i]._set(edge.perp().normalize())
+	}
+	aabb.Add(new /datum/point(min_x, max_y))
+	aabb.Add(new /datum/point(max_x, min_y))
+	aabb.Add(new /datum/point(min_x, min_y))
+	aabb.Add(new /datum/point(min_x, max_y))
+
+/**
+
+Simple method to calculate whether we collide with another shape object, lightweight but not hugely precise.
+
+*/
+
+/datum/shape/proc/test_aabb(var/datum/shape/other)
+	return ((src.aabb[1] + src.position.x) <= (other.aabb[3] + other.position.x)) && ((src.aabb[2] + src.position.y) <= (other.aabb[4] + other.position.y)) && ((src.aabb[3] + src.position.x) >= (other.aabb[1] + other.position.x)) && ((src.aabb[4] + src.position.y) >= (other.aabb[2] + other.position.y))
+
+/**
+
+Method to test if two shape objects are colliding with one another. We start off with a relatively inexpensive "Axis Oriented Bounding Box" check to test a block collision. If theyre not colliding with that, then it's safe
+to say that we don't need the added cost (and extra precision) of SAT.
+
+@returns boolean true / false
+
+*/
+
+/datum/shape/proc/collides(var/datum/shape/other)
+	if(!src.test_aabb(other)){
+		return FALSE
+	}
+	for (var/norm in src.normals){
+		if(is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```

Comment:
we want to return if they *are* colliding, which means *no* separating axis so it should be
```suggestion
		if(!is_separating_axis(src.position, other.position, src.rel_points, other.rel_points, norm)){
```